cmd/agent/main.go

```diff
@@ -44,86 +47,22 @@ func main() {
 	util_log.InitLogger(&cfg.Server)
 	logger := util_log.Logger
 
-	var (
-		promMetrics *prom.Agent
-		lokiLogs    *loki.Loki
-		tempoTraces *tempo.Tempo
-		manager     *integrations.Manager
-	)
-
-	srv, err := server.New(cfg.Server)
+	srv, err := NewEntryPoint(logger, cfg)
 	if err != nil {
-		level.Error(logger).Log("msg", "failed to create server", "err", err)
+		level.Error(logger).Log("msg", "error creating the agent server entrypoint", "err", err)
 		os.Exit(1)
 	}
-
-	if cfg.Prometheus.Enabled {
-		promMetrics, err = prom.New(prometheus.DefaultRegisterer, cfg.Prometheus, logger)
-		if err != nil {
-			level.Error(logger).Log("msg", "failed to create prometheus instance", "err", err)
-			os.Exit(1)
-		}
-
-		// Hook up API paths to the router
-		promMetrics.WireAPI(srv.HTTP)
-		promMetrics.WireGRPC(srv.GRPC)
-	}
-
-	if cfg.Loki.Enabled {
-		lokiLogs, err = loki.New(prometheus.DefaultRegisterer, cfg.Loki, logger)
-		if err != nil {
-			level.Error(logger).Log("msg", "failed to create loki log collection instance", "err", err)
-			os.Exit(1)
-		}
-	}
-
-	if cfg.Tempo.Enabled {
-		tempoTraces, err = tempo.New(prometheus.DefaultRegisterer, cfg.Tempo, cfg.Server.LogLevel)
-		if err != nil {
-			level.Error(logger).Log("msg", "failed to create tempo trace collection instance", "err", err)
-			os.Exit(1)
-		}
-	}
-
-	if cfg.Integrations.Enabled {
-		manager, err = integrations.NewManager(cfg.Integrations, logger, promMetrics.InstanceManager())
-		if err != nil {
-			level.Error(logger).Log("msg", "failed to create integrations manager", "err", err)
-			os.Exit(1)
-		}
-
-		if err := manager.WireAPI(srv.HTTP); err != nil {
-			level.Error(logger).Log("msg", "failed wiring endpoints for integrations", "err", err)
-			os.Exit(1)
-		}
-	}
-
-	srv.HTTP.HandleFunc("/-/healthy", func(w http.ResponseWriter, r *http.Request) {
-		w.WriteHeader(http.StatusOK)
-		fmt.Fprintf(w, "Agent is Healthy.\n")
-	})
-	srv.HTTP.HandleFunc("/-/ready", func(w http.ResponseWriter, r *http.Request) {
-		w.WriteHeader(http.StatusOK)
-		fmt.Fprintf(w, "Agent is Ready.\n")
-	})
-
-	if err := srv.Run(); err != nil {
+	exit := make(chan error)
+	go func() {
+		err := srv.Start()
+		exit <- err
+	}()
+	err = <-exit
+	if err != nil {
```

Comment:
You don't need the goroutine here because there's nothing else running concurrently, you can just call `srv.Start` directly